server/logger.go

```diff
@@ -8,69 +8,57 @@ import (
 	"net/http"
 	"runtime"
 	"strings"
-	"time"
 
 	"github.com/theplant/appkit/contexts"
-	"github.com/theplant/appkit/log"
+	"github.com/theplant/appkit/logtracing"
 )
 
-// Will absorb panics in earlier Middleware. Times the request and logs the result. FIXME split the timing out into a separate Middleware
+// Will absorb panics in earlier Middleware. Times the request and logs the result.
 func LogRequest(h http.Handler) http.Handler {
 	return http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
-		start := time.Now()
 		path := r.RequestURI
 
-		ctx := r.Context()
-
-		logger, ok := log.FromContext(ctx)
-
-		if !ok {
-			h.ServeHTTP(rw, r)
-			return
-		}
+		ctx, span := logtracing.StartSpan(r.Context(), fmt.Sprintf("%s %s", r.Method, path))
+		r = r.WithContext(ctx)
+		span.AppendKVs(
+			logtracing.HTTPServerKVs(r)...,
+		)
 
-		l := logger.With(
+		// NOTE for compatibility
+		span.AppendKVs(
 			"context", "http",
 			"path", path,
 			"method", r.Method,
 			"client_ip", clientIP(r),
 		)
 
-		l.Debug().Log("msg", fmt.Sprintf("%s %s", r.Method, path))
-
 		defer func() {
-			duration := int64(time.Since(start) / time.Microsecond)
-
 			status, _ := contexts.HTTPStatus(r.Context())
+			span.AppendKVs(
+				"http.status", status,
+			)
 
-			l = l.With(
-				"request_us", duration,
+			// NOTE for compatibility
+			span.End()
```

Comment:
Does this backwards compatibility cause the complexity with separating _ending_ spans and _logging_ spans? Because you want to find out the span duration and attach that as a KV before logging the span?

Reply:
> The span API looks a bit complex, is this mostly to support backwards compatibility for HTTP server logging?

As my thoughts, we can finish most of the tracing work with:
- `logtracing.StartSpan`
- `logtracing.AppendSpanKVs`
- `logtracing.EndSpan`

And in some cases, we can manipulate the span flexibly with:
- `(*span).AppendKVs`
- `(*span).End`
- `(*span).RecordError`
- `(*span).RecordPanic`
- `logtracing.LogSpan`

Reply:
`logtracing.EndSpan` = `(*span).RecordError` + `(*span).End` + `logtracing.LogSpan`

I think this separating can provide flexibility.

> Because you want to find out the span duration and attach that as a KV before logging the span?

The `LogReqeust` backward compatibility is not why I did the design. The design can resolve cases like this 🙃 (this is not a case I excepted...)

Reply:
Fine with me, thanks for the explanation.